### PR TITLE
Support external source Aerospike

### PR DESCRIPTION
**Pull request by Gleb-Tretyakov**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

For changelog. Remove if this is non-significant change.

Category (leave one):
- New Feature

Short description (up to few sentences):

Support aerospike storage as external dictionary

### CODE REVIEW

File: dbms/src/Dictionaries/AerospikeBlockInputStream.cpp [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+                    break;
+                case ValueType::Int32:
+                    static_cast<ColumnVector<Int32> &>(column).getData().push_back(key->value.integer.value);
+                    break;
+                case ValueType::Int64:
+                    static_cast<ColumnVector<Int64> &>(column).getData().push_back(key->value.integer.value);
+                    break;
+                case ValueType::String:
+                {
+                    String str{key->value.string.value, key->value.string.len};
```
Comment (comunodi) [inline dbms/src/Dictionaries/AerospikeBlockInputStream.cpp]: no need to create String

File: dbms/src/Dictionaries/AerospikeBlockInputStream.cpp [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+
+        size_t getNumRows() const { return num_rows; }
+
+    private:
+        void insertKey(IColumn & column, const ValueType type, const as_key * key, const std::string & name) const
+        {
+            switch (type)
+            {
+                case ValueType::UInt8:
+                    static_cast<ColumnVector<UInt8> &>(column).getData().push_back(key->value.integer.value);
```
Comment (comunodi) [inline dbms/src/Dictionaries/AerospikeBlockInputStream.cpp]: insertValue(...)

File: dbms/src/Dictionaries/AerospikeBlockInputStream.cpp [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+    TemporaryName recordsHandler(&columns, description, cursor);
+    for (UInt32 i = 0; i < current_block_size; ++i)
+    {
+        InitializeBatchKey(as_batch_keyat(&batch, i), "test", "test_set", keys[cursor + i]);
+    }
+
+    const auto batchReadCallback = [](const as_batch_read * results, uint32_t size, void * records_handler_)
+    {
+        TemporaryName * records_handler = static_cast<TemporaryName *>(records_handler_);
+        uint32_t n_found = 0;
```
Comment (comunodi) [inline dbms/src/Dictionaries/AerospikeBlockInputStream.cpp]: unused

File: dbms/src/Dictionaries/AerospikeBlockInputStream.cpp [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+                const as_bin_value & bin_value = record.bins.entries[idx - 1].value;
+                if (as_val_type(const_cast<as_bin_value*>(&bin_value)) == AS_NIL)
+                {
+                    insertDefaultValue(*(*columns)[idx], *description.sample_block.getByPosition(idx).column);
+                }
+                else
+                {
+                    if (description.types[idx].second)
+                    {
+                        ColumnNullable & column_nullable = static_cast<ColumnNullable &>(*(*columns)[idx]);
```
Comment (comunodi) [inline dbms/src/Dictionaries/AerospikeBlockInputStream.cpp]: cast to ColumnNullable* to remove double dereferencing here

File: dbms/src/Dictionaries/AerospikeBlockInputStream.cpp [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+    description.init(sample_block);
+}
+
+AerospikeBlockInputStream::~AerospikeBlockInputStream() = default;
+
+namespace
+{
+    using ValueType = ExternalResultDescription::ValueType;
+
+    class TemporaryName
```
Comment (comunodi) [inline dbms/src/Dictionaries/AerospikeBlockInputStream.cpp]: naming

File: dbms/src/Dictionaries/AerospikeBlockInputStream.cpp [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+        {
+            switch (as_val_type(const_cast<as_bin_value*>(value)))
+            {
+                case AS_INTEGER:
+                    static_cast<ColumnVector<T> &>(column).getData().push_back(value->integer.value);
+                    break;
+                case AS_DOUBLE:
+                    static_cast<ColumnVector<T> &>(column).getData().push_back(value->dbl.value);
+                    break;
+                    /*case AS_STRING:
```
Comment (comunodi) [inline dbms/src/Dictionaries/AerospikeBlockInputStream.cpp]: remove or implement

File: dbms/src/Dictionaries/AerospikeBlockInputStream.cpp [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+                else
+                {
+                    if (description.types[idx].second)
+                    {
+                        ColumnNullable & column_nullable = static_cast<ColumnNullable &>(*(*columns)[idx]);
+                        insertValue(column_nullable.getNestedColumn(), description.types[idx].first, &bin_value, name);
+                        column_nullable.getNullMapData().emplace_back(0);
+                    }
+                    else
+                        insertValue(*(*columns)[idx], description.types[idx].first, &bin_value, name);
```
Comment (comunodi) [inline dbms/src/Dictionaries/AerospikeBlockInputStream.cpp]: same here

File: dbms/src/Dictionaries/AerospikeBlockInputStream.cpp [hunk 8]
```diff
@@ -0,0 +1,337 @@
...
+            }
+        }
+
+        // Aerospike batch result return only bins. Keys must be processed separately
+        void HandleKeys(const std::vector<std::unique_ptr<as_key>> & keys) const
+        {
+            const auto & name = description.sample_block.getByPosition(0).name; // MAY BE MOVE TO CLASS FIELDS
+            for (const auto & key : keys)
+            {
+                insertKey(*((*columns)[0]), description.types[0].first, key.get(),
```
Comment (comunodi) [inline dbms/src/Dictionaries/AerospikeBlockInputStream.cpp]: same here

File: dbms/src/Dictionaries/AerospikeBlockInputStream.cpp [hunk 9]
```diff
@@ -0,0 +1,337 @@
...
+                    break;
+                case ValueType::String:
+                {
+                    String str{key->value.string.value, key->value.string.len};
+                    static_cast<ColumnString &>(column).insertDataWithTerminatingZero(str.data(), str.size() + 1);
+                    break;
+                }
+                case ValueType::Date:
+                    static_cast<ColumnUInt16 &>(column).getData().push_back(
+                        UInt16{DateLUT::instance().toDayNum(static_cast<Int64>(key->value.integer.value))});
```
Comment (comunodi) [inline dbms/src/Dictionaries/AerospikeBlockInputStream.cpp]: parse LocalDate

File: dbms/src/Dictionaries/AerospikeBlockInputStream.cpp [hunk 10]
```diff
@@ -0,0 +1,337 @@
...
+                    break;
+                case ValueType::UUID:
+                {
+                    String str{key->value.string.value, key->value.string.len};
+                    static_cast<ColumnUInt128 &>(column).getData().push_back(parse<UUID>(str));
+                    break;
+                }
+                default:
+                    std::string invalid_type = "bad"; // toString(as_val_type(&(key->value))); // TODO:FIX_ME(glebx777)
+                    throw Exception{"Type mismatch, expected String (UUID), got type id = " + invalid_type + " for column " + name,
```
Comment (comunodi) [inline dbms/src/Dictionaries/AerospikeBlockInputStream.cpp]: change message

File: dbms/src/Dictionaries/AerospikeBlockInputStream.cpp [hunk 11]
```diff
@@ -0,0 +1,337 @@
...
+                case ValueType::Int32:
+                    static_cast<ColumnVector<Int32> &>(column).getData().push_back(key->value.integer.value);
+                    break;
+                case ValueType::Int64:
+                    static_cast<ColumnVector<Int64> &>(column).getData().push_back(key->value.integer.value);
+                    break;
+                case ValueType::String:
+                {
+                    String str{key->value.string.value, key->value.string.len};
+                    static_cast<ColumnString &>(column).insertDataWithTerminatingZero(str.data(), str.size() + 1);
```
Comment (comunodi) [inline dbms/src/Dictionaries/AerospikeBlockInputStream.cpp]: parse String

File: dbms/src/Dictionaries/AerospikeBlockInputStream.h [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+#include <memory>
+#include <aerospike/aerospike.h>
+# include <aerospike/as_key.h>
+#include <Core/Block.h>
+#include <DataStreams/IBlockInputStream.h>
+#include <Core/ExternalResultDescription.h>
+
+namespace DB
+{
+/// Converts MongoDB Cursor to a stream of Blocks
```
Comment (comunodi) [inline dbms/src/Dictionaries/AerospikeBlockInputStream.h]: remove comment

File: dbms/src/Dictionaries/AerospikeDictionarySource.cpp [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+                                 Block & sample_block,
+                                 const Context & /* context */) -> DictionarySourcePtr {
+#if USE_AEROSPIKE
+        return std::make_unique<AerospikeDictionarySource>(dict_struct, config, config_prefix + ".aerospike", sample_block);
+#else
+        (void)dict_struct;
+        (void)config;
+        (void)config_prefix;
+        (void)sample_block;
+        throw Exception{"Dictionary source of type `aerospike` is disabled because poco library was built without aerospike support.",
```
Comment (comunodi) [inline dbms/src/Dictionaries/AerospikeDictionarySource.cpp]: remove mention of poco

File: dbms/src/Dictionaries/AerospikeDictionarySource.cpp [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+          dict_struct, config.getString(config_prefix + ".host"), config.getUInt(config_prefix + ".port"), nullptr, sample_block)
+{
+}
+
+AerospikeDictionarySource::AerospikeDictionarySource(const DB::AerospikeDictionarySource & other)
+    : dict_struct{other.dict_struct}, host{other.host}, port{other.port}, client{other.client}
+{
+}
+
+const static char * default_namespace_name = "test";
```
Comment (comunodi) [inline dbms/src/Dictionaries/AerospikeDictionarySource.cpp]: move to config

File: dbms/src/Dictionaries/AerospikeDictionarySource.cpp [hunk 15]
```diff
@@ -0,0 +1,167 @@
...
+    for (UInt64 id : ids)
+    {
+        keys.emplace_back(as_key_new_int64(default_namespace_name, default_set_name, id));
+    }
+    return std::make_shared<AerospikeBlockInputStream>(client, std::move(keys), sample_block, max_block_size);
+}
+
+std::string AerospikeDictionarySource::toString() const
+{
+    return "Aerospike: "; // +  bla bla bla
```
Comment (comunodi) [inline dbms/src/Dictionaries/AerospikeDictionarySource.cpp]: more info

File: dbms/src/Dictionaries/AerospikeDictionarySource.h [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+    BlockInputStreamPtr loadUpdatedAll() override
+    {
+        throw Exception{"Method loadUpdatedAll is unsupported for AerospikeDictionarySource", ErrorCodes::NOT_IMPLEMENTED};
+    }
+
+    bool supportsSelectiveLoad() const override { return true; }
+
+    /// @TODO(glebx777): fix it
+    BlockInputStreamPtr loadIds(const std::vector<UInt64> & /*ids*/) override;
+    /* override {
```
Comment (comunodi) [inline dbms/src/Dictionaries/AerospikeDictionarySource.h]: remove commented

File: dbms/src/Dictionaries/AerospikeDictionarySource.h [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+            throw Exception{"Method loadIds is unsupported for AerospikeDictionarySource", ErrorCodes::NOT_IMPLEMENTED};
+    } */
+
+    BlockInputStreamPtr loadKeys(const Columns & /* key_columns */, const std::vector<size_t> & /* requested_rows */) override
+    {
+            //Aerospke does not support native indexing
+            throw Exception{"Method loadKeys is unsupported for AerospikeDictionarySource", ErrorCodes::NOT_IMPLEMENTED};
+    }
+
+    /// @todo: for Aerospike, modification date can somehow be determined from the `_id` object field
```
Comment (comunodi) [inline dbms/src/Dictionaries/AerospikeDictionarySource.h]: fix comment

File: dbms/src/Dictionaries/AerospikeDictionarySource.h [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+    ~AerospikeDictionarySource() override;
+
+    BlockInputStreamPtr loadAll() override;
+
+    BlockInputStreamPtr loadUpdatedAll() override
+    {
+        throw Exception{"Method loadUpdatedAll is unsupported for AerospikeDictionarySource", ErrorCodes::NOT_IMPLEMENTED};
+    }
+
+    bool supportsSelectiveLoad() const override { return true; }
```
Comment (comunodi) [inline dbms/src/Dictionaries/AerospikeDictionarySource.h]: false?

File: cmake/find_aerospike.cmake [hunk 19]
```diff
@@ -0,0 +1,14 @@
...
+        if (NOT EXISTS "${ClickHouse_SOURCE_DIR}/contrib/aerospike")
+            message (WARNING "submodule contrib/aerospike is missing. to fix try run: \n git submodule update --init --recursive")
+        else()
+            set (AEROSPIKE_INCLUDE_DIR ${ClickHouse_SOURCE_DIR}/contrib/aerospike/src/include ${CMAKE_SOURCE_DIR}/contrib/aerospike/modules/common/src/include ${CMAKE_SOURCE_DIR}/contrib/aerospike/modules/mod-lua/src/include)
+            set (AEROSPIKE_LIBRARY aerospike)
+            set (USE_AEROSPIKE 1)
+        endif()
+    endif ()
+endif()
+message(STATUS "Using Aerospike=${USE_AEROSPIKE}: ${USE_POCO_REDIS}, ${AEROSPIKE_LIBRARY}, ${AEROSPIKE_INCLUDE_DIR}")
```
Comment (comunodi) [inline cmake/find_aerospike.cmake]: change vars

File: cmake/find_aerospike.cmake [hunk 20]
```diff
@@ -0,0 +1,14 @@
+if(OS_FREEBSD AND SANITIZE STREQUAL "address")
+    if (NOT EXISTS "${ClickHouse_SOURCE_DIR}/contrib/aerospike/src/main/aerospike/_bin.c")
```
Comment (comunodi) [inline cmake/find_aerospike.cmake]: + ENABLE_AEROSPIKE

File: cmake/find_aerospike.cmake [hunk 21]
```diff
@@ -0,0 +1,16 @@
+if (NOT DEFINED ENABLE_AEROSPIKE OR ENABLE_AEROSPIKE)
+    if(OS_FREEBSD AND SANITIZE STREQUAL "address")
```
Comment (4ertus2) [inline cmake/find_aerospike.cmake]: Do not build on Linux. `else ()` block missing

**Review by 4ertus2 — CHANGES_REQUESTED** [review]

File: cmake/find_aerospike.cmake (hunk 21 above)
Comment (4ertus2) [reply cmake/find_aerospike.cmake]: Fails to build with Aerospike enabled
```
../dbms/src/Dictionaries/AerospikeDictionarySource.cpp:173:5: error: no matching function for call to 'printf'
    printf(stderr, "len %zu", strlen(keys[0]->value.string.value));
    ^~~~~~
/usr/include/stdio.h:318:12: note: candidate function not viable: no known conversion from 'struct _IO_FILE *' to 'const char *__restrict' for 1st argument
extern int printf (const char *__restrict __format, ...);
           ^
1 error generated.
```

**Comment by abyss7** [comment]
Continued in #5629

**Pull request closed without merging**